eclipse-theia/theia
# Fix Git status issues in Source Control view

#### What it does

Fixes https://github.com/eclipse-theia/theia/issues/7789

Fixes issues with the the Source Control view.
- The list of staged and unstaged files are incorrect
- The status indicators are incorrect

I have attached here four different screen-shots of the same working tree and index.  The following actions were taken:
- FileA was created, staged, then modified
- FileB was created, staged, then deleted
- git-diff-contribution.ts was renamed to git-diff-contributionA.ts, staged, then modified

This is what we see with this PR:

![Screenshot from 2020-07-22 15-44-37](https://user-images.githubusercontent.com/87310/88191409-5209e500-cc33-11ea-9fd8-ef70c26a4ed6.png)

This is what we see in current Theia master.  FileB is missing altogether which is not correct because if the user were to commit then FileB would be committed.  FileA shows as Untracked in the working tree which is incorrect because the file has been staged and the user would expect to be able to see the changes in the file which have not been staged.  git-diff-contributionA.ts shows as a Rename in both groups but it should show as Modified in the unstaged section because it was not renamed again, just modified.

![Screenshot from 2020-07-22 15-58-12](https://user-images.githubusercontent.com/87310/88192085-0a378d80-cc34-11ea-909b-0286eff58c36.png)

This is what we see in VS-Code.  This is not correct.

![Screenshot from 2020-07-22 16-01-25](https://user-images.githubusercontent.com/87310/88192471-731f0580-cc34-11ea-8a47-e55d0f7ad4cc.png)

This is what we see in the git-gui tool (staged and unstaged sections are reversed).  This matches what we see with this PR except that the rename is shown as an add and a delete.  Selecting either of the two modified files in the unstaged changes will show the changes between the working tree and the index.

![Screenshot from 2020-07-22 15-55-19](https://user-images.githubusercontent.com/87310/88191709-a319d900-cc33-11ea-849c-faa1cd6f2cfa.png)

You will notice that this code calls dugite, not dugite-extra, for the status.  This change was made because the status function in dugite-extra has been written to support a simplified UI in which there is no Staged or Unstaged section shown but just a single list of files that can be checked (to include in the commit) or unchecked (to exclude from the commit).  This UI is used, for example, by Github Desktop.  Clients that show separate staged and unstaged sections should use the status as returned by git status.

The initial code to check the git version was copied from dugite-extra.  We could change dugite-extra to expose that code so it is not duplicated.  I don't know about the other clients of dugite-extra so I have not done that.
 
You will notice some of the diffs do not open.  For example if you rename a file and make a change, then stage it.  Selecting the file in the staged section, one would expect to see a diff with the change.  The problem is caused because GitScmProvider.getUriToOpen is not using the old uri for the HEAD uri.  This is changed by https://github.com/eclipse-theia/theia/pull/8084 so will be fixed when that is merged.

#### How to test

The three cases above are not exhaustive.  There are numerous combinations of Rename, Copy, Add, Modify, and Delete that can be in the staged and unstaged sections.  Ideally all these combinations should be tested.

#### Review checklist

- [x ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
